// @polkadot/api-derive/council/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  members: true,
  hasProposals: true,
  proposal: true,
  proposalCount: true,
  proposalHashes: true,
  proposals: true,
  prime: true
};
exports.proposals = exports.proposalHashes = exports.proposalCount = exports.proposal = exports.prime = exports.members = exports.hasProposals = void 0;

var _index = require("../collective/index.cjs");

var _votes = require("./votes.cjs");

Object.keys(_votes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _votes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _votes[key];
    }
  });
});

var _votesOf = require("./votesOf.cjs");

Object.keys(_votesOf).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _votesOf[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _votesOf[key];
    }
  });
});
// Copyright 2017-2022 @polkadot/api-derive authors & contributors
// SPDX-License-Identifier: Apache-2.0
const members = (0, _index.members)('council');
exports.members = members;
const hasProposals = (0, _index.hasProposals)('council');
exports.hasProposals = hasProposals;
const proposal = (0, _index.proposal)('council');
exports.proposal = proposal;
const proposalCount = (0, _index.proposalCount)('council');
exports.proposalCount = proposalCount;
const proposalHashes = (0, _index.proposalHashes)('council');
exports.proposalHashes = proposalHashes;
const proposals = (0, _index.proposals)('council');
exports.proposals = proposals;
const prime = (0, _index.prime)('council');
exports.prime = prime;
