// @polkadot/util/bi/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  nSqrt: true,
  nToBigInt: true,
  nMax: true,
  nMin: true,
  nToHex: true,
  nToU8a: true
};
Object.defineProperty(exports, "nMax", {
  enumerable: true,
  get: function () {
    return _min.nMax;
  }
});
Object.defineProperty(exports, "nMin", {
  enumerable: true,
  get: function () {
    return _min.nMin;
  }
});
Object.defineProperty(exports, "nSqrt", {
  enumerable: true,
  get: function () {
    return _sqrt.nSqrt;
  }
});
Object.defineProperty(exports, "nToBigInt", {
  enumerable: true,
  get: function () {
    return _toBigInt.nToBigInt;
  }
});
Object.defineProperty(exports, "nToHex", {
  enumerable: true,
  get: function () {
    return _toHex.nToHex;
  }
});
Object.defineProperty(exports, "nToU8a", {
  enumerable: true,
  get: function () {
    return _toU8a.nToU8a;
  }
});

var _consts = require("./consts.cjs");

Object.keys(_consts).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _consts[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _consts[key];
    }
  });
});

var _sqrt = require("./sqrt.cjs");

var _toBigInt = require("./toBigInt.cjs");

var _min = require("./min.cjs");

var _toHex = require("./toHex.cjs");

var _toU8a = require("./toU8a.cjs");
