// @polkadot/types-codec/primitive/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "I128", {
  enumerable: true,
  get: function () {
    return _I5.i128;
  }
});
Object.defineProperty(exports, "I16", {
  enumerable: true,
  get: function () {
    return _I2.i16;
  }
});
Object.defineProperty(exports, "I256", {
  enumerable: true,
  get: function () {
    return _I6.i256;
  }
});
Object.defineProperty(exports, "I32", {
  enumerable: true,
  get: function () {
    return _I3.i32;
  }
});
Object.defineProperty(exports, "I64", {
  enumerable: true,
  get: function () {
    return _I4.i64;
  }
});
Object.defineProperty(exports, "I8", {
  enumerable: true,
  get: function () {
    return _I.i8;
  }
});
Object.defineProperty(exports, "U128", {
  enumerable: true,
  get: function () {
    return _U5.u128;
  }
});
Object.defineProperty(exports, "U16", {
  enumerable: true,
  get: function () {
    return _U2.u16;
  }
});
Object.defineProperty(exports, "U256", {
  enumerable: true,
  get: function () {
    return _U6.u256;
  }
});
Object.defineProperty(exports, "U32", {
  enumerable: true,
  get: function () {
    return _U3.u32;
  }
});
Object.defineProperty(exports, "U64", {
  enumerable: true,
  get: function () {
    return _U4.u64;
  }
});
Object.defineProperty(exports, "U8", {
  enumerable: true,
  get: function () {
    return _U.u8;
  }
});
Object.defineProperty(exports, "USize", {
  enumerable: true,
  get: function () {
    return _USize.usize;
  }
});
Object.defineProperty(exports, "i128", {
  enumerable: true,
  get: function () {
    return _I5.i128;
  }
});
Object.defineProperty(exports, "i16", {
  enumerable: true,
  get: function () {
    return _I2.i16;
  }
});
Object.defineProperty(exports, "i256", {
  enumerable: true,
  get: function () {
    return _I6.i256;
  }
});
Object.defineProperty(exports, "i32", {
  enumerable: true,
  get: function () {
    return _I3.i32;
  }
});
Object.defineProperty(exports, "i64", {
  enumerable: true,
  get: function () {
    return _I4.i64;
  }
});
Object.defineProperty(exports, "i8", {
  enumerable: true,
  get: function () {
    return _I.i8;
  }
});
Object.defineProperty(exports, "u128", {
  enumerable: true,
  get: function () {
    return _U5.u128;
  }
});
Object.defineProperty(exports, "u16", {
  enumerable: true,
  get: function () {
    return _U2.u16;
  }
});
Object.defineProperty(exports, "u256", {
  enumerable: true,
  get: function () {
    return _U6.u256;
  }
});
Object.defineProperty(exports, "u32", {
  enumerable: true,
  get: function () {
    return _U3.u32;
  }
});
Object.defineProperty(exports, "u64", {
  enumerable: true,
  get: function () {
    return _U4.u64;
  }
});
Object.defineProperty(exports, "u8", {
  enumerable: true,
  get: function () {
    return _U.u8;
  }
});
Object.defineProperty(exports, "usize", {
  enumerable: true,
  get: function () {
    return _USize.usize;
  }
});

var _I = require("./I8.cjs");

var _I2 = require("./I16.cjs");

var _I3 = require("./I32.cjs");

var _I4 = require("./I64.cjs");

var _I5 = require("./I128.cjs");

var _I6 = require("./I256.cjs");

var _U = require("./U8.cjs");

var _U2 = require("./U16.cjs");

var _U3 = require("./U32.cjs");

var _U4 = require("./U64.cjs");

var _U5 = require("./U128.cjs");

var _U6 = require("./U256.cjs");

var _USize = require("./USize.cjs");
