// @polkadot/api-derive/technicalCommittee/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.proposals = exports.proposalHashes = exports.proposalCount = exports.proposal = exports.prime = exports.members = exports.hasProposals = void 0;

var _index = require("../collective/index.cjs");

// Copyright 2017-2022 @polkadot/api-derive authors & contributors
// SPDX-License-Identifier: Apache-2.0
const members = (0, _index.members)('technicalCommittee');
exports.members = members;
const hasProposals = (0, _index.hasProposals)('technicalCommittee');
exports.hasProposals = hasProposals;
const proposal = (0, _index.proposal)('technicalCommittee');
exports.proposal = proposal;
const proposalCount = (0, _index.proposalCount)('technicalCommittee');
exports.proposalCount = proposalCount;
const proposalHashes = (0, _index.proposalHashes)('technicalCommittee');
exports.proposalHashes = proposalHashes;
const proposals = (0, _index.proposals)('technicalCommittee');
exports.proposals = proposals;
const prime = (0, _index.prime)('technicalCommittee');
exports.prime = prime;
