// @polkadot/util-crypto/ed25519/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "convertPublicKeyToCurve25519", {
  enumerable: true,
  get: function () {
    return _convertKey.convertPublicKeyToCurve25519;
  }
});
Object.defineProperty(exports, "convertSecretKeyToCurve25519", {
  enumerable: true,
  get: function () {
    return _convertKey.convertSecretKeyToCurve25519;
  }
});
Object.defineProperty(exports, "ed25519DeriveHard", {
  enumerable: true,
  get: function () {
    return _deriveHard.ed25519DeriveHard;
  }
});
Object.defineProperty(exports, "ed25519PairFromRandom", {
  enumerable: true,
  get: function () {
    return _fromRandom.ed25519PairFromRandom;
  }
});
Object.defineProperty(exports, "ed25519PairFromSecret", {
  enumerable: true,
  get: function () {
    return _fromSecret.ed25519PairFromSecret;
  }
});
Object.defineProperty(exports, "ed25519PairFromSeed", {
  enumerable: true,
  get: function () {
    return _fromSeed.ed25519PairFromSeed;
  }
});
Object.defineProperty(exports, "ed25519PairFromString", {
  enumerable: true,
  get: function () {
    return _fromString.ed25519PairFromString;
  }
});
Object.defineProperty(exports, "ed25519Sign", {
  enumerable: true,
  get: function () {
    return _sign.ed25519Sign;
  }
});
Object.defineProperty(exports, "ed25519Verify", {
  enumerable: true,
  get: function () {
    return _verify.ed25519Verify;
  }
});

var _convertKey = require("./convertKey.cjs");

var _deriveHard = require("./deriveHard.cjs");

var _fromRandom = require("./pair/fromRandom.cjs");

var _fromSecret = require("./pair/fromSecret.cjs");

var _fromSeed = require("./pair/fromSeed.cjs");

var _fromString = require("./pair/fromString.cjs");

var _sign = require("./sign.cjs");

var _verify = require("./verify.cjs");
