// @polkadot/types-codec/base/Compact.js
// Copyright 2017-2022 @polkadot/types-codec authors & contributors
// SPDX-License-Identifier: Apache-2.0
import { compactFromU8a, compactToU8a, isBigInt, isBn, isNumber, isString } from '@polkadot/util';
import { typeToConstructor } from "../utils/index.js";
/**
 * @name Compact
 * @description
 * A compact length-encoding codec wrapper. It performs the same function as Length, however
 * differs in that it uses a variable number of bytes to do the actual encoding. This is mostly
 * used by other types to add length-prefixed encoding, or in the case of wrapped types, taking
 * a number and making the compact representation thereof
 */

export class Compact {
  #Type;
  #raw;

  constructor(registry, Type, value = 0) {
    this.registry = registry;
    this.#Type = typeToConstructor(registry, Type);
    const [raw, decodedLength] = Compact.decodeCompact(registry, this.#Type, value);
    this.initialU8aLength = decodedLength;
    this.#raw = raw;
  }

  static with(Type) {
    return class extends Compact {
      constructor(registry, value) {
        super(registry, Type, value);
      }

    };
  }
  /** @internal */


  static decodeCompact(registry, Type, value) {
    if (value instanceof Compact) {
      return [new Type(registry, value.#raw), 0];
    } else if (value instanceof Type) {
      return [value, 0];
    } else if (isString(value) || isNumber(value) || isBn(value) || isBigInt(value)) {
      return [new Type(registry, value), 0];
    }

    const [decodedLength, bn] = compactFromU8a(value);
    return [new Type(registry, bn), decodedLength];
  }
  /**
   * @description The length of the value when encoded as a Uint8Array
   */


  get encodedLength() {
    return this.toU8a().length;
  }
  /**
   * @description returns a hash of the contents
   */


  get hash() {
    return this.registry.hash(this.toU8a());
  }
  /**
   * @description Checks if the value is an empty value
   */


  get isEmpty() {
    return this.#raw.isEmpty;
  }
  /**
   * @description Returns the number of bits in the value
   */


  bitLength() {
    return this.#raw.bitLength();
  }
  /**
   * @description Compares the value of the input to see if there is a match
   */


  eq(other) {
    return this.#raw.eq(other instanceof Compact ? other.#raw : other);
  }
  /**
   * @description Returns a breakdown of the hex encoding for this Codec
   */


  inspect() {
    return {
      outer: [this.toU8a()]
    };
  }
  /**
   * @description Returns a BigInt representation of the number
   */


  toBigInt() {
    return this.#raw.toBigInt();
  }
  /**
   * @description Returns the BN representation of the number
   */


  toBn() {
    return this.#raw.toBn();
  }
  /**
   * @description Returns a hex string representation of the value. isLe returns a LE (number-only) representation
   */


  toHex(isLe) {
    return this.#raw.toHex(isLe);
  }
  /**
   * @description Converts the Object to to a human-friendly JSON, with additional fields, expansion and formatting of information
   */


  toHuman(isExtended) {
    return this.#raw.toHuman(isExtended);
  }
  /**
   * @description Converts the Object to JSON, typically used for RPC transfers
   */


  toJSON() {
    return this.#raw.toJSON();
  }
  /**
   * @description Returns the number representation for the value
   */


  toNumber() {
    return this.#raw.toNumber();
  }
  /**
   * @description Returns the base runtime type name for this instance
   */


  toRawType() {
    return `Compact<${this.registry.getClassName(this.#Type) || this.#raw.toRawType()}>`;
  }
  /**
   * @description Returns the string representation of the value
   */


  toString() {
    return this.#raw.toString();
  }
  /**
   * @description Encodes the value as a Uint8Array as per the SCALE specifications
   * @param isBare true when the value has none of the type-specific prefixes (internal)
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars


  toU8a(isBare) {
    return compactToU8a(this.#raw.toBn());
  }
  /**
   * @description Returns the embedded [[UInt]] or [[Moment]] value
   */


  unwrap() {
    return this.#raw;
  }

}
