// @polkadot/util/string/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "stringCamelCase", {
  enumerable: true,
  get: function () {
    return _camelCase.stringCamelCase;
  }
});
Object.defineProperty(exports, "stringLowerFirst", {
  enumerable: true,
  get: function () {
    return _lowerFirst.stringLowerFirst;
  }
});
Object.defineProperty(exports, "stringPascalCase", {
  enumerable: true,
  get: function () {
    return _camelCase.stringPascalCase;
  }
});
Object.defineProperty(exports, "stringShorten", {
  enumerable: true,
  get: function () {
    return _shorten.stringShorten;
  }
});
Object.defineProperty(exports, "stringToHex", {
  enumerable: true,
  get: function () {
    return _toHex.stringToHex;
  }
});
Object.defineProperty(exports, "stringToU8a", {
  enumerable: true,
  get: function () {
    return _toU8a.stringToU8a;
  }
});
Object.defineProperty(exports, "stringUpperFirst", {
  enumerable: true,
  get: function () {
    return _lowerFirst.stringUpperFirst;
  }
});

var _camelCase = require("./camelCase.cjs");

var _lowerFirst = require("./lowerFirst.cjs");

var _shorten = require("./shorten.cjs");

var _toHex = require("./toHex.cjs");

var _toU8a = require("./toU8a.cjs");
